modules/metrics/src/main/java/org/apache/ignite/internal/metrics/MetricRegistry.java

```diff
@@ -0,0 +1,300 @@
+    public void disable(@NotNull MetricSource src) {
+        modifySources(sources -> {
+            checkRegistered(sources, src);
+
+            src.disable();
+
+            IgniteBiTuple<MetricSource, MetricSet> updated = new IgniteBiTuple<>(src, null);
+
+            sources.put(src.name(), updated);
+
+            return true;
+        });
+    }
+
+    /**
+     * Disable metric set for the given metric source.
+     *
+     * @param srcName Metric source name.
+     * @throws IllegalStateException If metric source with given name doesn't exists.
+     */
+    public void disable(final String srcName) {
+        modifySources(sources -> {
+            IgniteBiTuple<MetricSource, MetricSet> registered = sources.get(srcName);
+
+            if (registered == null) {
+                throw new IllegalStateException("Metrics source with given name doesn't exists: " + srcName);
+            }
+
+            MetricSource src = registered.get1();
+
+            src.disable();
+
+            IgniteBiTuple<MetricSource, MetricSet> updated = new IgniteBiTuple<>(src, null);
+
+            sources.put(src.name(), updated);
+
+            return true;
+        });
+    }
+
+    /**
+     * Check that the given metric source is registered.
+     *
+     * @param sources Sources map.
+     * @param src Metric source.
+     * @return Registered pair of metric source and metric set.
+     * @throws IllegalStateException If metric source isn't registered.
+     * @throws IllegalArgumentException If metric source isn't the same as registered.
+     */
+    @NotNull
+    private IgniteBiTuple<MetricSource, MetricSet> checkRegistered(
+            Map<String, IgniteBiTuple<MetricSource, MetricSet>> sources,
+            @NotNull MetricSource src
+    ) {
+        requireNonNull(src);
+
+        IgniteBiTuple<MetricSource, MetricSet> registered = sources.get(src.name());
+
+        if (registered == null) {
+            throw new IllegalStateException("Metrics source isn't registered: " + src.name());
+        }
+
+        if (!src.equals(registered.get1())) {
+            throw new IllegalArgumentException("Given metric source is not the same as registered by the same name: " + src.name());
+        }
+
+        return registered;
+    }
+
+    /**
+     * Updates {@link MetricRegistry#sources} map using copy-on-write principle. Increments version of registry.
+     *
+     * @param modifier Modifier for the sources map. Accepts the new, modifiable version of the map. Returns boolean value, whether
+     *                 the map was modified.
+     */
+    private void modifySources(Function<Map<String, IgniteBiTuple<MetricSource, MetricSet>>, Boolean> modifier) {
+        lock.lock();
+
+        try {
+            Map<String, IgniteBiTuple<MetricSource, MetricSet>> sources0 = new TreeMap<>(sources);
+
+            boolean modified = modifier.apply(sources0);
+
+            if (modified) {
+                sources = sources0;
+
+                version++;
```

Comment:
It is incorrect behavior because, for example, registering of metric source doesn't change the metrics schema until will not be enabled.

The previous implementation was better because metrics are versioned while sources are not.

Reply:
I fixed this and added the tests for checking the correctness of version. Registering the new source seems to be the only case when the map changes and version doesnt, and the old implementation didnt provide copy-on-write semantics so I will leave the current one, in general.